File: hedera-mirror-common/src/test/java/com/hedera/mirror/common/domain/transaction/TransactionTest.java
/*
 * Copyright (C) 2023-2025 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.mirror.common.domain.transaction;

import static com.hedera.mirror.common.converter.ObjectToStringSerializer.OBJECT_MAPPER;
import static org.assertj.core.api.Assertions.assertThat;

import com.hedera.mirror.common.domain.DomainBuilder;
import com.hedera.mirror.common.domain.entity.EntityId;
import com.hedera.mirror.common.domain.token.NftTransfer;
import org.junit.jupiter.api.Test;
import org.skyscreamer.jsonassert.JSONAssert;
import org.skyscreamer.jsonassert.JSONCompareMode;

class TransactionTest {

    private static final String EXPECTED_JSON_TEMPLATE =
            """
                    {
                      "consensus_timestamp": 1684791152000000000,
                      "charged_tx_fee": 1,
                      "entity_id": 2,
                      "errata": "INSERT",
                      "index":4,
                      "initial_balance": 5,
                      "itemized_transfer": %s,
                      "max_custom_fees": %s,
                      "memo": "BgcI",
                      "max_fee": 9,
                      "nft_transfer": %s,
                      "node_account_id": 3,
                      "nonce": 19,
                      "parent_consensus_timestamp": 20,
                      "payer_account_id": 21,
                      "result": 22,
                      "scheduled": false,
                      "transaction_bytes": "FxgZ",
                      "transaction_hash": "Ghsc",
                      "transaction_record_bytes": "HR4f",
                      "type": 29,
                      "valid_duration_seconds": 30,
                      "valid_start_ns": 31
                    }
                    """;
    private static final String EXPECTED_MAX_CUSTOM_FEES = "[\"AQI=\", \"Cgs=\"]";
    private static final String EXPECTED_ITEMIZED_TRANSFER_VALUE =
            """
                    "[{\\"amount\\":-200,\\"entity_id\\":50,\\"is_approval\\":true},{\\"amount\\":200,\\"entity_id\\":51,\\"is_approval\\":false}]"
                    """;
    private static final String EXPECTED_NFT_TRANSFER_VALUE =
            """
                    "[{\\"is_approval\\":false,\\"receiver_account_id\\":10,\\"sender_account_id\\":11,\\"serial_number\\":12,\\"token_id\\":13},{\\"is_approval\\":true,\\"receiver_account_id\\":14,\\"sender_account_id\\":15,\\"serial_number\\":16,\\"token_id\\":17}]"
                    """;

    @Test
    void addItemizedTransfer() {
        var transaction = Transaction.builder().build();
        assertThat(transaction.getItemizedTransfer()).isNull();

        var domainBuilder = new DomainBuilder();
        var itemizedTransfer1 = ItemizedTransfer.builder()
                .amount(domainBuilder.number())
                .entityId(domainBuilder.entityId())
                .isApproval(false)
                .build();
        transaction.addItemizedTransfer(itemizedTransfer1);
        assertThat(transaction.getItemizedTransfer()).containsExactly(itemizedTransfer1);

        var itemizedTransfer2 = ItemizedTransfer.builder()
                .amount(domainBuilder.number())
                .entityId(domainBuilder.entityId())
                .isApproval(true)
                .build();
        transaction.addItemizedTransfer(itemizedTransfer2);
        assertThat(transaction.getItemizedTransfer()).containsExactly(itemizedTransfer1, itemizedTransfer2);
    }

    @Test
    void addNftTransfer() {
        var transaction = Transaction.builder().build();
        assertThat(transaction.getNftTransfer()).isNull();

        var domainBuilder = new DomainBuilder();
        var nftTransfer1 = domainBuilder.nftTransfer().get();
        transaction.addNftTransfer(nftTransfer1);
        assertThat(transaction.getNftTransfer()).containsExactly(nftTransfer1);

        var nftTransfer2 = domainBuilder.nftTransfer().get();
        transaction.addNftTransfer(nftTransfer2);
        assertThat(transaction.getNftTransfer()).containsExactly(nftTransfer1, nftTransfer2);
    }

    @Test
    void toJson() throws Exception {
        // given
        var transaction = getTransaction();
        var itemizedTransfer1 = ItemizedTransfer.builder()
                .amount(-200L)
                .entityId(EntityId.of(50))
                .isApproval(true)
                .build();
        transaction.addItemizedTransfer(itemizedTransfer1);

        var itemizedTransfer2 = ItemizedTransfer.builder()
                .amount(200L)
                .entityId(EntityId.of(51))
                .isApproval(false)
                .build();
        transaction.addItemizedTransfer(itemizedTransfer2);

        var nftTransfer1 = new NftTransfer();
        nftTransfer1.setIsApproval(false);
        nftTransfer1.setReceiverAccountId(EntityId.of(10L));
        nftTransfer1.setSenderAccountId(EntityId.of(11L));
        nftTransfer1.setSerialNumber(12L);
        nftTransfer1.setTokenId(EntityId.of(13L));
        transaction.addNftTransfer(nftTransfer1);

        NftTransfer nftTransfer2 = new NftTransfer();
        nftTransfer2.setIsApproval(true);
        nftTransfer2.setReceiverAccountId(EntityId.of(14L));
        nftTransfer2.setSenderAccountId(EntityId.of(15L));
        nftTransfer2.setSerialNumber(16L);
        nftTransfer2.setTokenId(EntityId.of(17L));
        transaction.addNftTransfer(nftTransfer2);

        // when
        String actual = OBJECT_MAPPER.writeValueAsString(transaction);

        // then
        String expected = String.format(
                EXPECTED_JSON_TEMPLATE,
                EXPECTED_ITEMIZED_TRANSFER_VALUE,
                EXPECTED_MAX_CUSTOM_FEES,
                EXPECTED_NFT_TRANSFER_VALUE);
        JSONAssert.assertEquals(expected, actual, JSONCompareMode.STRICT);
    }

    @Test
    void toJsonNullItemizedTransferAndNullMaxCustomFeesAndNullNftTransfer() throws Exception {
        // given
        var transaction = getTransaction();
        transaction.setMaxCustomFees(null);

        // when
        String actual = OBJECT_MAPPER.writeValueAsString(transaction);

        // then
        String expected = String.format(EXPECTED_JSON_TEMPLATE, "null", "null", "null");
        JSONAssert.assertEquals(expected, actual, JSONCompareMode.STRICT);
    }

    private Transaction getTransaction() {
        var transaction = new Transaction();
        transaction.setConsensusTimestamp(1684791152000000000L);
        transaction.setChargedTxFee(1L);
        transaction.setEntityId(EntityId.of(2L));
        transaction.setErrata(ErrataType.INSERT);
        transaction.setIndex(4);
        transaction.setInitialBalance(5L);
        transaction.setMaxCustomFees(new byte[][] {{0x1, 0x2}, {0xa, 0xb}});
        transaction.setMemo(new byte[] {6, 7, 8});
        transaction.setMaxFee(9L);
        transaction.setNodeAccountId(EntityId.of(3L));
        transaction.setNonce(19);
        transaction.setParentConsensusTimestamp(20L);
        transaction.setPayerAccountId(EntityId.of(21L));
        transaction.setResult(22);
        transaction.setScheduled(false);
        transaction.setTransactionBytes(new byte[] {23, 24, 25});
        transaction.setTransactionHash(new byte[] {26, 27, 28});
        transaction.setTransactionRecordBytes(new byte[] {29, 30, 31});
        transaction.setType(29);
        transaction.setValidDurationSeconds(30L);
        transaction.setValidStartNs(31L);
        return transaction;
    }
}

File: hedera-mirror-rosetta/app/persistence/domain/transaction_test.go
/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

package domain

import (
	"fmt"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestTransactionTableName(t *testing.T) {
	assert.Equal(t, "transaction", Transaction{}.TableName())
}

func TestItemizedTransferSliceScan(t *testing.T) {
	// given
	bytes := []byte(`[
{"amount": -20, "entity_id": 101, "is_approval": true},
{"amount": 10, "entity_id": 102, "is_approval": false},
{"amount": 10, "entity_id": 103, "is_approval": false}
]`)
	expected := ItemizedTransferSlice{
		{Amount: -20, EntityId: EntityId{EntityNum: 101, EncodedId: 101}, IsApproval: true},
		{Amount: 10, EntityId: EntityId{EntityNum: 102, EncodedId: 102}},
		{Amount: 10, EntityId: EntityId{EntityNum: 103, EncodedId: 103}},
	}

	// when
	actual := ItemizedTransferSlice{}
	err := actual.Scan(bytes)

	// then
	assert.NoError(t, err)
	assert.Equal(t, expected, actual)
}

func TestItemizedTransferSliceScanThrows(t *testing.T) {
	inputs := []interface{}{
		"not byte array",
		[]byte(`[{"amount": -20, "entity_id": 101, "is_approval": true]`), // invalid json string
	}

	for _, input := range inputs {
		t.Run(fmt.Sprintf("%s", input), func(t *testing.T) {
			actual := ItemizedTransferSlice{}
			err := actual.Scan(input)
			assert.Error(t, err)
		})
	}
}

func TestItemizedTransferSliceValue(t *testing.T) {
	// given
	transfer := ItemizedTransferSlice{
		{Amount: -20, EntityId: EntityId{EntityNum: 101, EncodedId: 101}, IsApproval: true},
		{Amount: 10, EntityId: EntityId{EntityNum: 102, EncodedId: 102}},
		{Amount: 10, EntityId: EntityId{EntityNum: 103, EncodedId: 103}},
	}
	expected := `[
{"amount":-20, "entity_id": 101, "is_approval": true},
{"amount":10, "entity_id": 102, "is_approval": false},
{"amount":10, "entity_id": 103, "is_approval": false}
]`
	expected = strings.ReplaceAll(strings.ReplaceAll(expected, " ", ""), "\n", "")

	// when
	actual, err := transfer.Value()

	// then
	assert.NoError(t, err)
	assert.IsType(t, []byte{}, actual)
	bytes, _ := actual.([]byte)
	assert.Equal(t, expected, string(bytes))
}

func TestItemizedTransferSliceValueEmpty(t *testing.T) {
	actual, err := ItemizedTransferSlice{}.Value()
	assert.NoError(t, err)
	assert.Nil(t, actual)
}
